watch/watch.go

```diff
@@ -0,0 +1,360 @@
+//
+// WatchQuery will return an error if registering the watch fails for any reason.
+func (w *Watcher) Query(query, id string) (<-chan Event, error) {
+	out := make(chan Event)
+	if err := w.query(query, id, out); err != nil {
+		close(out)
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// EndQuery ends the watch on the query associated with id. It will close the channel
+// that was delivering notifications for the given ID. This may happen before or after
+// EndQuery returns.
+func (w *Watcher) EndQuery(id string) error {
+	w.l.Lock()
+	defer w.l.Unlock()
+
+	return w.endQuery(id, false)
+}
+
+// Close ends the watches on all queries this Watcher has. If closing the Watcher fails,
+// no changes will be made to it.
+//
+// Further attempts to register or end watches will result in an error after Close() is
+// called.
+func (w *Watcher) Close(txn storage.Transaction) error {
+	w.l.Lock()
+	defer w.l.Unlock()
+
+	if w.closed {
+		return errors.New("watcher already closed")
+	}
+
+	w.close(txn, false)
+	return nil
+}
+
+// Migrate creates a new Watcher with the same watches as w, but using the new
+// ast.Compiler. Like when creating a Watcher with New, the provided compiler must not
+// be modified after being passed to Migrate, or else behavior is undefined.
+//
+// Migrate will not start the new Watcher, it must be started with the Start function
+// before it will deliver notifications.
+//
+// After Migrate returns, the old watcher will be closed, and the new will be ready for
+// use. All channels from the old watcher will still be active, via the returned Watcher.
+//
+// If an error occurs creating the new Watcher, the state of the old Watcher will not be
+// changed.
+func (w *Watcher) Migrate(c *ast.Compiler, txn storage.Transaction) (*Watcher, error) {
+	w.l.Lock()
+	defer w.l.Unlock()
+	if w.closed {
+		return nil, errors.New("watcher is closed")
+	}
+
+	m := create(w.ctx, w.store, c, w.id)
+	for id, notify := range w.queryWatch {
+		if err := m.query(notify.query, id, notify.out); err != nil {
+			return nil, err
+		}
+	}
+
+	w.close(txn, true)
+
+	// Don't start the new watcher. This allows us to properly guarantee that
+	// the old watcher will not be changed if setting up the new one fails. By making
+	// starting the watcher the client's responsibility, we don't have to worry about
+	// failures in the store preventing it from registering a trigger.
+	return m, nil
+}
+
+func create(ctx context.Context, s storage.Store, c *ast.Compiler, id string) *Watcher {
+	w := &Watcher{
+		id:       id,
+		store:    s,
+		compiler: c,
+		ctx:      ctx,
+
+		queryWatch: map[string]queryNotify{},
+		dataWatch:  map[string]map[signal]struct{}{},
+	}
+	w.indexRules(c)
+	return w
+}
+
+func (w *Watcher) query(query, id string, out chan<- Event) error {
+	w.l.Lock()
+	defer w.l.Unlock()
+
+	if w.closed {
+		return errors.New("watcher is closed")
+	}
+
+	if _, ok := w.queryWatch[id]; ok {
+		return fmt.Errorf("already a watch with id %s", id)
```

Comment:
If the Watcher is not split up per comment above, then I'd consider making the query implementation idempotent.